### PR TITLE
Trusted keys

### PR DESCRIPTION
**Pull request by Max-Siegieda**
This allows the gnupg module to accept a "trust_level" argument which then creates an exported ownertrust file with the key fingerprint and trust level before importing it. You can use this to make keys on a machine be ultimately trusted.

### CODE REVIEW

**Comment by mattj-io** [comment]
I have no idea what any of this means or does, but it looks legit :+1:

File: lib/puppet/provider/gnupg_key/gnupg.rb [hunk 1]
```diff
@@ -73,7 +79,7 @@ def add_key_from_key_content
     path = create_temporary_file(user_id, resource[:key_content])
     command = "gpg --import #{path}"
     begin
-      output = Puppet::Util::Execution.execute(command, :uid => user_id, :failonfail => true)
+      `#{command}`
```
Comment (bilco105) [inline lib/puppet/provider/gnupg_key/gnupg.rb]:
```
Puppet::Util::Execution.execute()
```

File: lib/puppet/provider/gnupg_key/gnupg.rb [hunk 2]
```diff
@@ -136,6 +142,20 @@ def puppet_content
...
 
+  def create_trustdb
+    fingerprint = get_fingerprint_from_key_id
+    fingerprint = fingerprint.delete!("\n")
+    ownertrust = fingerprint + ":#{resource[:trust_level]}:\n"
+    path = create_temporary_file(user_id, ownertrust)
+
+    command = "gpg --import-ownertrust #{path}"
+    begin
+      `#{command}`
```
Comment (bilco105) [inline lib/puppet/provider/gnupg_key/gnupg.rb]:
```
Puppet::Util::Execution.execute()
```

File: lib/puppet/type/gnupg_key.rb [hunk 3]
```diff
@@ -139,6 +139,18 @@
     end
   end
 
+  newparam(:trust_level) do
+    desc "Trust level. This is the gpg owner trust level for the key."
+
+    validate do |value|
+      unless value =~ /^[1-6]$/
+        raise ArgumentError, "Invalid trust level"
```
Comment (bilco105) [inline lib/puppet/type/gnupg_key.rb]:
``` ruby
raise(ArgumentError, "Invalid trust level") unless value =~ /^[1-6]$/
```

File: lib/puppet/provider/gnupg_key/gnupg.rb (hunk 1 above)
Comment (Max-Siegieda) [inline lib/puppet/provider/gnupg_key/gnupg.rb]: That raises an exception and halts the script despite everything being fine, literally output of gpg includes no errors, exception gets thrown anyway. This way puppet only errors if there's... you know, an error.

TL;DR rabbit hooooole

Comment (spjmurray) [inline lib/puppet/provider/gnupg_key/gnupg.rb]: so, what's the return code?  also the rescue will die with an undefined variable now

File: lib/puppet/provider/gnupg_key/gnupg.rb [hunk 4]
```diff
@@ -50,6 +45,17 @@ def add_key
...
       add_key_from_key_content
     end
+    create_trustdb
+  end
+
+  def get_fingerprint_from_key_id
+    begin
+      fingerprint_command = "gpg --fingerprint --with-colons #{resource[:key_id]} | awk -F: '$1 == \"fpr\" {print $10;}'"
+      fingerprint = Puppet::Util::Execution.execute(fingerprint_command, :uid => user_id)
+      return fingerprint
```
Comment (spjmurray) [inline lib/puppet/provider/gnupg_key/gnupg.rb]: fyi you don't need the return, gives away the last expression for free

Comment (seanhandley) [inline lib/puppet/provider/gnupg_key/gnupg.rb]: Agreed. You don't even need the variable. Just

``` ruby
begin
  fingerprint_command = "gpg --fingerprint --with-colons #{resource[:key_id]} | awk -F: '$1 == \"fpr\" {print $10;}'"
  Puppet::Util::Execution.execute(fingerprint_command, :uid => user_id)
rescue
 # ...
end
```

File: lib/puppet/provider/gnupg_key/gnupg.rb (hunk 1 above)
Comment (seanhandley) [inline lib/puppet/provider/gnupg_key/gnupg.rb]: Yeah, you need to catch the returned error code and raise it as a puppet error i.e.

```ruby
code = `foo`
raise(Puppet::Error, "Failed with code #{code}") unless code == "0"

Comment (Max-Siegieda) [inline lib/puppet/provider/gnupg_key/gnupg.rb]: Return code... well, tacking this on the end doesn't still results in no output. Running things manually I get no such problems and the output code is 0. _digs further_

```
command = "gpg --import #{path} && echo $?"
```

File: lib/puppet/provider/gnupg_key/gnupg.rb [hunk 5]
```diff
@@ -50,6 +45,16 @@ def add_key
...
       add_key_from_key_content
     end
+    create_trustdb
+  end
+
+  def get_fingerprint_from_key_id
+    begin
+      fingerprint_command = "gpg --fingerprint --with-colons #{resource[:key_id]} | awk -F: '$1 == \"fpr\" {print $10;}'"
+      Puppet::Util::Execution.execute(fingerprint_command, :uid => user_id)
+    rescue Puppet::ExecutionFailure => e
```
Comment (seanhandley) [inline lib/puppet/provider/gnupg_key/gnupg.rb]:
```ruby
def fingerprint
  @fingerprint ||= ...
end
```

File: lib/puppet/provider/gnupg_key/gnupg.rb [hunk 6]
```diff
@@ -50,6 +45,16 @@ def add_key
...
+      raise Puppet::Error, "Could not determine fingerprint for  #{resource[:key_id]} for user #{resource[:user]}: #{fingerprint}"
+    end
   end
```
Comment (seanhandley) [inline lib/puppet/provider/gnupg_key/gnupg.rb:47]:
``` ruby
  def fingerprint
    @fingerprint ||= begin
      fingerprint_command = "gpg --fingerprint --with-colons #{resource[:key_id]} | awk -F: '$1 == \"fpr\" {print $10;}'"
      Puppet::Util::Execution.execute(fingerprint_command, :uid => user_id)
    rescue Puppet::ExecutionFailure => e
      raise Puppet::Error, "Could not determine fingerprint for  #{resource[:key_id]} for user #{resource[:user]}: #{fingerprint}"
    end
   end
```

**Comment by yankcrime** [comment]
:+1: 

¯_(ツ)_/¯ If @seanhandley and @bilco105 are happy then saul goodman.

File: lib/puppet/provider/gnupg_key/gnupg.rb [hunk 7]
```diff
@@ -73,9 +76,9 @@ def add_key_from_key_content
     path = create_temporary_file(user_id, resource[:key_content])
     command = "gpg --import #{path}"
     begin
-      output = Puppet::Util::Execution.execute(command, :uid => user_id, :failonfail => true)
+    output = Puppet::Util::Execution.execute(command, :uid => user_id, :failonfail => true)
     rescue Puppet::ExecutionFailure => e
-      raise Puppet::Error, "Error while importing key #{resource[:key_id]} using key content:\n#{output}}"
+      raise Puppet::Error, "Error while importing key #{resource[:key_id]} using key content: #{output}"
```
Comment (seanhandley) [inline lib/puppet/provider/gnupg_key/gnupg.rb]: If puppet raises an exception during the execute here, the `output` variable isn't going to get set.

This means that when you raise a `Puppet::Error` in the rescue-block, the `output` variable is going to be `nil`. That won't lead to an unhandled exception, but it means Ruby will just interpolate nothing into the error string.

I think you want:

``` ruby
raise Puppet::Error, "Error while importing key #{resource[:key_id]} using key content: #{resource[:key_content]}"
```

File: lib/puppet/provider/gnupg_key/gnupg.rb [hunk 8]
```diff
@@ -136,6 +139,18 @@ def puppet_content
     @content = tmp.content
   end
 
+  def create_trustdb
+    ownertrust = fingerprint.delete("\n") + ":#{resource[:trust_level]}:\n"
+    path = create_temporary_file(user_id, ownertrust)
+
+    command = "gpg --import-ownertrust #{path}"
+    begin
+      output = Puppet::Util::Execution.execute(command, :uid => user_id, :failonfail => true)
```
Comment (seanhandley) [inline lib/puppet/provider/gnupg_key/gnupg.rb]: Since you're never using the output variable, there's no point setting it here.

File: lib/puppet/type/gnupg_key.rb [hunk 9]
```diff
@@ -146,4 +156,4 @@
 
     defaultto :public
   end
-end
+end
```
Comment (seanhandley) [inline lib/puppet/type/gnupg_key.rb]: Newline at the end of the file FTW

File: lib/puppet/provider/gnupg_key/gnupg.rb (hunk 7 above)
Comment (seanhandley) [inline lib/puppet/provider/gnupg_key/gnupg.rb]: If you want to get the reason for the `Puppet::ExecutionFailure` then you can probably check the `e` variable. Try calling `#inspect` on it, then you can work that into the `Puppet::Error` you're raising, thereby helping people figure out why it failed.

Comment (Max-Siegieda) [inline lib/puppet/provider/gnupg_key/gnupg.rb]: I actually want the output of the command in case of failure, and that explains why I was getting nothing when trying to debug it earlier. what do Sean what do?

File: lib/puppet/provider/gnupg_key/gnupg.rb [hunk 10]
```diff
@@ -32,12 +25,7 @@ def remove_key
     elsif resource[:key_type] == :both
       command = "gpg --batch --yes --delete-secret-and-public-key #{fingerprint}"
     end
```
Comment (seanhandley) [inline lib/puppet/provider/gnupg_key/gnupg.rb:27]: http://stackoverflow.com/questions/948135/how-can-i-write-a-switch-statement-in-ruby

File: lib/puppet/provider/gnupg_key/gnupg.rb [hunk 11]
```diff
@@ -136,6 +114,14 @@ def puppet_content
     @content = tmp.content
   end
 
+  def create_trustdb
+    ownertrust = fingerprint.delete("\n") + ":#{resource[:trust_level]}:\n"
+    path = create_temporary_file(user_id, ownertrust)
+
+    command = "gpg --import-ownertrust #{path}"
+    run_gpg_command(command)
```
Comment (seanhandley) [inline lib/puppet/provider/gnupg_key/gnupg.rb:122]: For a short line like this, I'd just do

``` ruby
run_gpg_command("gpg --import-ownertrust #{path}")
```

**Comment by seanhandley** [comment]
:+1:
